DEPR offsets: rename 'Q' to 'QE'

natmokval opened the pull request:
xref #9586, #52064

Deprecated the alias denoting quarter end frequency `"Q"` for offsets in favour of `"QE"`.
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests)
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/v2.2.0.rst`

natmokval committed:
add dictionary OFFSET_DEPR_FREQSTR, change warning msg

natmokval committed:
rename Q to QE for offsets

natmokval committed:
resolve conflicts

natmokval committed:
correct def _adjust_bin_edges, dicts OFFSET_DEPR_FREQSTR, OFFSET_TO_PERIOD_FREQSTR, fix tests

natmokval committed:
correct def convert_json_field_to_pandas_type

natmokval committed:
correct doc

natmokval committed:
add tests, correct user_guide/timeseries.rst

natmokval committed:
fix pylint failure

natmokval committed:
correct docstring

natmokval committed:
Merge branch 'main' into rename-Q-QE-for-offsets

natmokval committed:
add a note to v2.2.0.rst

natmokval commented:
@MarcoGorelli, could you please take a look at this PR? It's a follow-up of PR: "DEPR offsets: rename 'M' to 'ME'".
seems like ci failure isn't related to my changes

MarcoGorelli commented on pandas/tests/indexes/datetimes/methods/test_to_period.py:
```diff
@@ -89,6 +89,23 @@ def test_dti_to_period_2monthish(self, freq_offset, freq_period):
+
+        rng = date_range("01-Jan-2012", periods=8, freq=freq)
+        prng = rng.to_period()
+        with tm.assert_produces_warning(UserWarning, match=msg):
```
Looks like I'd missed this last time, but this should be FutureWarning, not UserWarning https://github.com/pandas-dev/pandas/pull/52064#pullrequestreview-1690296955

could we address that in a separate PR please?

natmokval replied:
thank you, I corrected Warning class and made the separate PR

MarcoGorelli commented on pandas/tests/indexes/period/test_period_range.py:
```diff
@@ -152,3 +160,6 @@ def test_period_range_frequency_ME_error_message(self):
         with pytest.raises(ValueError, match=msg):
             period_range(start="Jan-2000", end="Dec-2000", freq="2ME")
+
+    def test_my(self):
+        Period("2017Q1", freq="Q")
```
temporary test name?😉

MarcoGorelli commented on pandas/tests/scalar/period/test_period.py:
```diff
@@ -73,10 +73,8 @@ def test_construction(self):
 
         i1 = Period.now(freq="Q")
         i2 = Period(datetime.now(), freq="Q")
-        i3 = Period.now("q")
```
could you remind me why this is removed please?

MarcoGorelli replied:
as is it just because we're not uppercasing qe? if so, sounds fine tbh

natmokval replied:
we had a strange behaviour, the same ass for deprecation M to ME:

```
Period('2020-01', 'QE') => errors
Period('2020-01', 'qe') => silently converts to 'M'
Period('2020-01', 'q') => errors
```
I added `'qe' `to `_dont_uppercase` and it fixed the problem. I left a comment linking to #54105, saying that this should raise in the future an error.

natmokval committed:
remove test_my

natmokval committed:
resolve conflicts

natmokval committed:
fix tests

MarcoGorelli commented on doc/source/whatsnew/v2.2.0.rst:
```diff
@@ -232,6 +232,28 @@ For example:
 
     pd.date_range('2020-01-01', periods=3, freq='ME')
 
+Deprecate alias ``Q`` in favour of ``QE`` for offsets
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
This looks very similar to the previous section, perhaps we can combine them?

natmokval replied:
yes, I agree. I combined these two sections together.

MarcoGorelli commented on pandas/io/json/_table_schema.py:
```diff
@@ -207,6 +208,8 @@ def convert_json_field_to_pandas_type(field) -> str | CategoricalDtype:
         if field.get("tz"):
             return f"datetime64[ns, {field['tz']}]"
         elif field.get("freq"):
+            # GH#9586 rename frequency M to ME for offsets
+            field["freq"] = freq_to_period_freqstr(1, field["freq"])
```
left over from https://github.com/pandas-dev/pandas/pull/55581#pullrequestreview-1693046244? should we wait for that one to be resolved first?

natmokval replied:
yes, it' related to #55581. I agree, it's better at first to merge #55581. I rolled my changes in `pandas/io/json/_table_schema.py` back , but now I have failures in `test_read_json_table_period_orient`, which will be resolved after merging.

MarcoGorelli reviewed (commented):
looks good, just this part isn't totally clear to me https://github.com/pandas-dev/pandas/pull/55553/files#r1367138011

natmokval committed:
roll back my changes in convert_json_field_to_pandas_type, correct v2.2.0.rst

natmokval committed:
add 'qe' to _dont_uppercase

MarcoGorelli reviewed (approved):
cool, should be good on green!

MarcoGorelli commented:
looks like there's some test failures
```
 FAILED pandas/tests/io/json/test_json_table_schema.py::TestTableOrientReader::test_read_json_table_period_orient[vals0-None] - TypeError: data type 'period[QE-DEC]' not understood
```

natmokval committed:
Merge branch 'main' into rename-Q-QE-for-offsets

natmokval committed:
add parameters Q, 2Q to test_read_json_table_orient_period_depr_freq

natmokval commented:
> looks like there's some test failures
> 
> ```
>  FAILED pandas/tests/io/json/test_json_table_schema.py::TestTableOrientReader::test_read_json_table_period_orient[vals0-None] - TypeError: data type 'period[QE-DEC]' not understood
> ```

`test_read_json_table_period_orient` failed because there was a bug for conversion JSON field descriptor into pandas type for deprecated offsets frequency 'M'. We fixed it in #55581, now the test passes.

I added parameters `'Q'` and `'2Q'` into parametrization of `test_read_json_table_orient_period_depr_freq`

natmokval committed:
correct DatetimeProperties docs

MarcoGorelli commented:
thanks @natmokval !
